### DISCUSSION TITLE
Show CircularProgressIndicator() on "loading" only after a minimum time

### DISCUSSION THREAD

**Discussion by herna** [rrousselGit/riverpod#3036] [Q&A]
Hi,

I have a typical AsyncNotifier and widget to perform some async operations.

When doing these async operations, like updating something in my data base, I emit an "AsyncValue.loading()" in the notifier to show a CircularProgressIndicator() and after that I perform the operation itself inside an "AsyncValue.guard()". In the build() method of my widget I have something like this:

```
Widget build(BuildContext context) {
  return MyWidget(
        body: myState.when(
            data: (data) => WhenDataWidget(data),
            error: (o, s) => WhenErrorWidget(oError: o),
            loading: () => const Center(child: CircularProgressIndicator()),
      );
}
```

As per the title of this post, I'd like to only show the CircularProgressIndicator() if the async operation takes a minimum time to avoid that weird flash if the operation only takes some milliseconds.

At the moment I am able to do it by using the timeout() function. I have these functions in my AsyncNotifier:

```
Future<void> asyncOp(Future<void> Function() fn) async {
    fn().timeout(const Duration(seconds: 1),
        onTimeout: () => state = const AsyncValue.loading());
  }

Future<void> updateData() async {
    state = await AsyncValue.guard(() async {
      // perform some operations and update the state
      return state.value!;
    });
  }

```

And then in the widget I trigger the async operation like this:

```
ref
  .read(myProvider().notifier).asyncOp(() => ref
    .read(myProvider().notifier).updateData());
```

This works but the problem is I don't know how to do it for all my providers without writing the asyncOp() in every of them.

Does Riverpod have some built-in functionality for this that I'm missing? How do you normally handle this?

Thank you.

**Comment by herna** [comment]
Hi @rrousselGit, could you just quickly let me know about this?

Very appreciated.

**Comment by AhmedLSayed9** [comment]
You can avoid the loading flash by setting `skipLoadingOnReload` to true at .when

**Comment by herna** [comment]
Thanks @AhmedLSayed9 for your reply.

If I set that `skipLoadingOnReload` to true, then the `loading: () => const Center(child: CircularProgressIndicator()),` is never shown while executing the async operation, that's not what I want but only showing this `CircularProgressIndicator()` if the operation takes longer than a particular time.

**Reply by AhmedLSayed9** [reply]
That's a known behavior "shown old data while the new data is loading".

If you want to show loading for a minimum time, there's no built-in way. The best you can do is:
```dart
extension FutureErrorX<T> on Future<T> {
  Future<void> minFuture([Duration duration = const Duration(seconds: 1)]) async {
    await (this, Future<void>.delayed(duration)).wait;
  }
}

// At your method
Future<void> updateData() async {
  state = await AsyncValue.guard(() async {
    await someFuture.minFuture();
    return state.value!;
  });
}
```

This will only delay your future when it's less than 1 second.

**Reply by herna** [reply]
Well, this is different from what I'm looking for, and what I am doing at the moment through `timeout()`

Your solution always waits a minimum time to perform the operation and hence delays that async operation that otherwise could be completed much faster.

`timeout()` is better in this case because it doesn't affect the operation itselft: if the operation takes less than the limit time (for example 1 second), then nothing is shown on the screen. If it takes longer than 1 second, then the `CircularProgressIndicator()` is shown and when the operation completes, no matter how long it takes, then the final result is shown.

My original question was if `Riverpod` had something to deal with this out of the box but it doesn't as far I can see, this is more Dart related I guess so I'll try to use `timeout()` the best way I can.

Thank you very much for your help.

**Reply by AhmedLSayed9** [reply]
> Your solution always waits a minimum time to perform the operation and hence delays that async operation that otherwise could be completed much faster.

It executes them in-parallel.

Alright. I see what you're looking for but your solution is not ideal as it'll stop waiting for the future ` fn()` after [timeLimit] has passed, which means you can't properly use the value returned from that future to mutate state.

Alternatively, you can use a mixin and a timer:
```dart
mixin GuardWithMinLoading<T> on AutoDisposeAsyncNotifier<T> {
  Future<AsyncValue<T>> guardWithMinLoading(Future<T> Function() future) async {
    try {
      final timer = Timer(const Duration(seconds: 1), () {
        state = const AsyncLoading();
      });
      final result = await future().whenComplete(timer.cancel);
      return AsyncValue.data(result);
    } catch (err, stack) {
      return AsyncValue.error(err, stack);
    }
  }
}


// At your notifier just add the mixin and do:
Future<void> updateData() async {
    state = await guardWithMinLoading(() async {
      final newData = await operation();
      return newData;
    });
  }
```

**Reply by herna** [reply]
Thank you very much @AhmedLSayed9, this seems to be pretty much what I was looking for!

I’ll give it a try.

**Reply by davidmartos96** [reply]
@herna Interesting idea. How are you handling the 4th initial empty state with AsyncValue.
Are you using the Data state with null?

**Reply by herna** [reply]
Hi @davidmartos96, are you talking about this? --> https://github.com/rrousselGit/riverpod/issues/2579

I have something like the following in every part of my application.

The widget:

```
final state =
        ref.watch(installationEditControllerProvider(widget.installationId));

return switch (state) {
      AsyncData(:final value, isLoading: false) => whenData(value),
      AsyncError(:final error) => WhenErrorWidget(oError: error),
      _ => const WhenLoadingWidget(),
    };
```

And the notifier:

```
@override
  FutureOr<InstallationEditState> build(int installationId) async {
    InstallationEditDTO installationEditDTO =
        await ref.read(installationServiceProvider).edit(installationId);
    return InstallationEditState(
      installationEditDTO: installationEditDTO,
    );
  }
```

The workflow of my application is: `build()` retrieves something from a database or an external API and I show the final page according to this data (almost always a form or a list or records)

I am only interested in showing the final page when the provider `build()` method completes, I don't have the need to make anything special for that initial empty state if that's what you mean. I am even using `isLoading: false` in `AsyncData` because sometimes I am invalidating the provider from other point of the application and I don't want the previous data to be shown, I just want the loading state in that case too.

I don't know if I have answered your question.

**Reply by davidmartos96** [reply]
@herna Thanks! Yes, that answers it. 
I was thinking you also didn't want to show the loading indicator with the first state, the initial build.

**Reply by herna** [reply]
Hi @davidmartos96 

Sorry but I'm still learning about Riverpod and didn't understand your question. I do really understand now that I was applying this solution and indeed I am stuck at how to apply this for that initial empty `build()`. Any ideas? I was trying to figure it out but I don't come up with a good solution.

@AhmedLSayed9, any help with this?

Thanks.
